### ISSUE TITLE
Could you mind to make a explanation for trainning GammaPoisson(alphas, betas) ?

### ISSUE THREAD

**Issue by eromoe** [FlorianWilhelm/bhm-at-scale#1]
Hello, first thank you to share the article , it is awesome !

I am trying to design a 4~5 level hierachical model now , and have a question about the parameter of `GammaPoisson` 

From 
https://github.com/FlorianWilhelm/bhm-at-scale/blob/2d6f9fc94b021034d16abfdc9f8d6fb1c0e242b1/src/bhm_at_scale/model.py#L148

```
means = (is_observed * jnp.exp(jnp.sum(jnp.expand_dims(coefs, axis=1) * features, axis=2))
         + not_observed*eps)

betas = is_observed*jnp.exp(-disp_params) + not_observed
alphas = means * betas
return numpyro.sample("days", dist.GammaPoisson(alphas, betas), obs=jnp.nan_to_num(targets))
```

I know such linear model called GLM (genelized linear model ) but I don't quite understand why you construct `alphas` and `betas`  in this way    .
Let me remove the `is_observed` and `not_observed` , aim to just focus `alphas` and `betas`  :
```
betas  = exp(-disp_params)  = 1/exp(disp_params)
alphas  = exp( sum(coefs * feature) -disp_params  ) = exp(sum(coefs * feature)/exp(disp_params)
```

I just can't understand why they are contructed like this .  I have viewed wiki and some sites to undertand negative binomial distribution and gamma-poisson distribution , but  doesn't help  to figure this out .

**Comment by FlorianWilhelm** [comment]
Hi @eromoe, thanks, great that you like my article.

Regarding your question, that's just a reparametrization I came up with, I haven't taken it from any book. I wanted that like in the case of the Poisson model the mean is given by `μ = exp(c^t * x)`, which is a generalized linear model like you stated.

In the article, I stated that for the variance we have  `σ2 = μ + 1/r*μˆ2` (check out [Models for Count Data With Overdispersion](https://data.princeton.edu/wws509/notes/c4a.pdf)). Now just looking at the [Wikipedia article of the Gamma Distribution](https://en.wikipedia.org/wiki/Gamma_distribution) (as I stated that NB is a Poisson distribution with `λ` drawn from a Gamma distribution), we have:

1. `μ =α/β`
2. `σ2 = α / βˆ2`

Since we defined `μ` and `r`, one could now use `σ2 = μ + 1/r*μˆ2` to calculate the variance. Using 1. and 2. one could then calculate `β = μ/σ2`.  I took a little shortcut here since I don't care about the "real" dispersion factors. With `exp(-disp_params)` I made sure that it's a positive number as `disp_params`is drawn from a Normal distribution and used it directly as `betas`. This saves some computations and should also make it easier for SGD. Have you tried to do it by the book?

**Comment by FlorianWilhelm** [comment]
Hi **[@eromoe](https://github.com/eromoe)**, thanks, great that you like my article.

Regarding your question, that's just a reformulation I came up with, I haven't taken it from any book. I wanted that like in the case of the Poisson model the mean is given by `μ = exp(c^t * x)`, which is a generalized linear model like you stated.

In the article, I stated that for the variance we have `σ² = μ + 1/r μ²` (check out [Models for Count Data With Overdispersion](https://data.princeton.edu/wws509/notes/c4a.pdf)). Now looking at `μ` and `σ²` of the Gamma-Poisson distribution from [here](http://www.math.wm.edu/~leemis/chart/UDR/PDFs/Gammapoisson.pdf) and noting that `β` should be `1/β` if the [original parametrization of the Gamma distribution](https://en.wikipedia.org/wiki/Gamma_distribution) was used, we have

1. `μ = α/β`
2. `σ² = α/β + α²/β = μ + βμ²`

Just comparing 2. to the variance above, we see that `β=1/r` and `α=βμ`. Since I needed to make sure that `r`, drawn from a Normal Distribution, is positive, I took `jnp.exp(-disp_params)`. I am not 100% sure though, so let me know if you think I am wrong.

The most important thing is that the `μ` is defined by the generalized linear model though and another parameter is used to adjust the variance.

**Comment by eromoe** [comment]
Thank you very much for detailed explanation !

**Comment by snowdustdj** [comment]
Thank you @FlorianWilhelm for the explanation. I am also interested in this topic. I am thinking if I want `r` to be positive, I can assign some distribution like half-Cauchy. My problem is much simpler than your HBM case, where I am only trying to fit a simple NB regression with some covariates (the following part only has intercept, trying to test my code since I'm new to Pyro). The MCMC sampler works fine but the SVI produces bunch of NAs. Not sure why that happened but I will try your way of paramaterization in a simulation to verify it is working.

```
mu = 0.3
variance = 1.6
phi = mu**2 / (variance-mu)
p = phi/(mu + phi)

y = np.random.negative_binomial(n=phi, p=p, size=1000000)

def model(y):
    b0 = pyro.sample("b0", dist.Normal(0, 5))
    phi = pyro.sample("phi", dist.HalfCauchy(2))
    mu = torch.exp(b0)
    beta = phi/mu
    alpha = phi
    with pyro.plate("data", len(y)):
        pyro.sample("obs", dist.GammaPoisson(alpha, beta), obs=y)

# MCMC works
from pyro.infer import MCMC, NUTS


nuts_kernel = NUTS(model)

mcmc = MCMC(nuts_kernel, num_samples=1000, warmup_steps=200)
y_tensor = torch.tensor(y, dtype=torch.float)
mcmc.run(y_tensor)

# the following fails
def guide(y):
    a_loc = pyro.param('a_loc', torch.tensor(0.))
    a_scale = pyro.param('a_scale', torch.tensor(1.),
                         constraint=constraints.positive)
    
    phi_loc = pyro.param('phi_loc', torch.tensor(1.),
                             constraint=constraints.positive)
    
    b0 = pyro.sample("b0", dist.Normal(a_loc, a_scale))
    phi = pyro.sample("phi", dist.Normal(phi_loc, torch.tensor(1.0)))

    mu = torch.exp(b0)
    beta = phi/mu
    alpha = phi

from pyro.infer import SVI, Trace_ELBO


svi = SVI(model,
          guide,
          optim.Adam({"lr": .01, "betas": (0.9, 0.999)}),
          loss=Trace_ELBO())

pyro.clear_param_store()
num_iters = 10000
loss = []
for i in range(num_iters):
    elbo = svi.step(y_tensor)
    loss.append(elbo)
    if i % 500 == 0:
        print("Elbo loss: {}".format(elbo))
```

**Comment by FlorianWilhelm** [comment]
If MCMC works then I would first look at the initialization of the guide, also maybe decrease the learning rate.
